File: src/tools/shared/configFiles.ts
import * as fs from "node:fs";
import * as path from "node:path";
import type { ServiceSpec } from "src/types/stack";

export interface BindMount {
  source: string;
  target: string;
  mode?: string;
}

export interface StagedConfigFile {
  path: string; // normalized, forward-slash, cwd-relative
  content: string;
  bytes: number;
}

const RESERVED_DIR = ".docker-agent";

/** Normalize a host-relative path to forward slashes with no leading "./". */
export function normalizeRel(p: string): string {
  const norm = path.normalize(p).split(path.sep).join("/");
  return norm.replace(/^\.\//, "");
}

/** Parse short-syntax volume "SOURCE:TARGET[:MODE]". Returns null for named volumes. */
export function parseBindMount(volume: string): BindMount | null {
  const parts = volume.split(":");
  if (parts.length < 2) return null;
  const source = parts[0] as string;
  const target = parts[1] as string;
  const mode = parts[2];
  // A bind mount's source is a filesystem path; named volumes start with an
  // alphanumeric. Windows absolute paths (C:\...) are out of scope.
  if (!/^[.~/]/.test(source)) return null;
  return mode ? { source, target, mode } : { source, target };
}

/** A bind source is "file-like" when its basename carries an extension. */
export function isFileLikeBind(source: string): boolean {
  return /\.[A-Za-z0-9]+$/.test(path.basename(source));
}

/** Confine a host-relative path to cwd. Rejects abs paths, traversal, reserved dirs. */
export function resolveSafe(
  cwd: string,
  relPath: string,
): { ok: true; abs: string } | { ok: false; error: string } {
  if (path.isAbsolute(relPath)) return { ok: false, error: "absolute paths are not allowed" };
  const abs = path.resolve(cwd, relPath);
  const rel = path.relative(cwd, abs);
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    return { ok: false, error: "path escapes the project directory" };
  }
  const first = normalizeRel(rel).split("/")[0];
  if (first === RESERVED_DIR) return { ok: false, error: `${RESERVED_DIR} is reserved` };
  return { ok: true, abs };
}

/** File-like bind mounts that have neither provided content nor an existing host file. */
export function detectMissingConfigFiles(
  services: Record<string, ServiceSpec>,
  providedKeys: Set<string>,
  cwd: string,
): Array<{ service: string; path: string }> {
  const missing: Array<{ service: string; path: string }> = [];
  for (const [service, spec] of Object.entries(services)) {
    for (const vol of spec.volumes ?? []) {
      const bind = parseBindMount(vol);
      if (!bind || !isFileLikeBind(bind.source)) continue;
      if (providedKeys.has(normalizeRel(bind.source))) continue;
      const safe = resolveSafe(cwd, bind.source);
      if (!safe.ok) continue; // unsafe paths are rejected during staging
      if (!fs.existsSync(safe.abs)) missing.push({ service, path: bind.source });
    }
  }
  return missing;
}

/**
 * File-like bind sources that would make `compose up` misbehave: the source is
 * absent (Docker silently auto-creates an empty *directory* there) or already
 * squats as a directory (mounting a dir onto an image file fails with "are you
 * trying to mount a directory onto a file"). Call this right before `up` so we
 * refuse loudly instead of letting Docker create a stray folder.
 */
export function findInvalidFileBinds(
  services: Record<string, ServiceSpec>,
  cwd: string,
): Array<{ service: string; path: string; reason: "missing" | "directory" }> {
  const bad: Array<{ service: string; path: string; reason: "missing" | "directory" }> = [];
  for (const [service, spec] of Object.entries(services)) {
    for (const vol of spec.volumes ?? []) {
      const bind = parseBindMount(vol);
      if (!bind || !isFileLikeBind(bind.source)) continue;
      const safe = resolveSafe(cwd, bind.source);
      if (!safe.ok) continue; // unsafe paths are rejected during staging
      if (!fs.existsSync(safe.abs)) {
        bad.push({ service, path: bind.source, reason: "missing" });
      } else if (fs.statSync(safe.abs).isDirectory()) {
        bad.push({ service, path: bind.source, reason: "directory" });
      }
    }
  }
  return bad;
}

const MAX_FILE_BYTES = 64 * 1024;
const MAX_TOTAL_BYTES = 256 * 1024;

export interface ConfigFileSnapshot {
  abs: string;
  existed: boolean;
  previousContent: string | null;
}

/** Validate provided configFiles against the services' file binds + size caps. */
export function stageConfigFiles(
  cwd: string,
  services: Record<string, ServiceSpec>,
  configFiles: Record<string, string> | undefined,
): { ok: true; staged: StagedConfigFile[] } | { ok: false; error: string } {
  const provided = configFiles ?? {};
  const fileBinds = new Set<string>();
  for (const spec of Object.values(services)) {
    for (const vol of spec.volumes ?? []) {
      const bind = parseBindMount(vol);
      if (bind && isFileLikeBind(bind.source)) fileBinds.add(normalizeRel(bind.source));
    }
  }
  let total = 0;
  const staged: StagedConfigFile[] = [];
  for (const [key, content] of Object.entries(provided)) {
    const safe = resolveSafe(cwd, key);
    if (!safe.ok) return { ok: false, error: `unsafe config file path "${key}": ${safe.error}` };
    const norm = normalizeRel(key);
    if (!fileBinds.has(norm)) {
      return { ok: false, error: `configFiles entry "${key}" matches no file bind mount` };
    }
    const bytes = Buffer.byteLength(content, "utf8");
    if (bytes > MAX_FILE_BYTES) return { ok: false, error: `config file "${key}" exceeds 64 KiB` };
    total += bytes;
    staged.push({ path: norm, content, bytes });
  }
  if (total > MAX_TOTAL_BYTES) return { ok: false, error: "config files total exceeds 256 KiB" };
  return { ok: true, staged };
}

export function snapshotConfigFiles(cwd: string, files: StagedConfigFile[]): ConfigFileSnapshot[] {
  return files.map((f) => {
    const abs = path.resolve(cwd, f.path);
    const existed = fs.existsSync(abs) && fs.statSync(abs).isFile();
    return {
      abs,
      existed,
      previousContent: existed ? fs.readFileSync(abs, "utf8") : null,
    };
  });
}

export function writeConfigFiles(cwd: string, files: StagedConfigFile[]): void {
  for (const f of files) {
    const safe = resolveSafe(cwd, f.path);
    if (!safe.ok) throw new Error(`refusing to write "${f.path}": ${safe.error}`);
    // Docker auto-creates an empty *directory* at a bind-mount source when the
    // file is missing at `compose up` time. If such a stale dir squats at our
    // target, remove it so writeFileSync lands a real file instead of throwing
    // EISDIR — and so the next mount sees a file, not a directory.
    if (fs.existsSync(safe.abs) && fs.statSync(safe.abs).isDirectory()) {
      fs.rmSync(safe.abs, { recursive: true, force: true });
    }
    fs.mkdirSync(path.dirname(safe.abs), { recursive: true });
    fs.writeFileSync(safe.abs, f.content, "utf8");
  }
}

export function restoreConfigFiles(snapshots: ConfigFileSnapshot[]): void {
  for (const s of snapshots) {
    if (s.existed) {
      fs.writeFileSync(s.abs, s.previousContent ?? "", "utf8");
    } else if (fs.existsSync(s.abs)) {
      // recursive covers a Docker-auto-created directory squatting at the path
      fs.rmSync(s.abs, { recursive: true, force: true });
    }
  }
}

File: servers/docker-mcp-server/src/docker_mcp_server/tools/shared/config_files.py
"""Config file bind-mount staging and rollback helpers.

Parity: ``src/tools/shared/configFiles.ts``.
"""

from __future__ import annotations

import os
import re
import shutil
from dataclasses import dataclass
from pathlib import Path

from docker_mcp_server.types.stack import ServiceSpec

RESERVED_DIR = ".docker-agent"
MAX_FILE_BYTES = 64 * 1024
MAX_TOTAL_BYTES = 256 * 1024


@dataclass
class BindMount:
    source: str
    target: str
    mode: str | None = None


@dataclass
class StagedConfigFile:
    path: str
    content: str
    bytes: int


@dataclass
class ConfigFileSnapshot:
    abs: str
    existed: bool
    previous_content: str | None


def normalize_rel(path: str) -> str:
    """Normalize a host-relative path to forward slashes with no leading './'."""
    norm = Path(path).as_posix()
    return re.sub(r"^\./", "", norm)


def parse_bind_mount(volume: str) -> BindMount | None:
    """Parse short-syntax volume ``SOURCE:TARGET[:MODE]``."""
    parts = volume.split(":")
    if len(parts) < 2:
        return None
    source = parts[0]
    target = parts[1]
    mode = parts[2] if len(parts) > 2 else None
    if not re.match(r"^[.~/]", source):
        return None
    return BindMount(source=source, target=target, mode=mode)


def is_file_like_bind(source: str) -> bool:
    """True when the host path basename carries an extension."""
    return bool(re.search(r"\.[A-Za-z0-9]+$", Path(source).name))


def resolve_safe(
    cwd: str | os.PathLike[str], rel_path: str
) -> dict[str, str] | dict[str, object]:
    """Confine a host-relative path to cwd."""
    if Path(rel_path).is_absolute():
        return {"ok": False, "error": "absolute paths are not allowed"}
    abs_path = Path(cwd).resolve() / rel_path
    abs_path = abs_path.resolve()
    cwd_resolved = Path(cwd).resolve()
    try:
        rel = abs_path.relative_to(cwd_resolved)
    except ValueError:
        return {"ok": False, "error": "path escapes the project directory"}
    first = normalize_rel(str(rel)).split("/")[0]
    if first == RESERVED_DIR:
        return {"ok": False, "error": f"{RESERVED_DIR} is reserved"}
    return {"ok": True, "abs": str(abs_path)}


def detect_missing_config_files(
    services: dict[str, ServiceSpec],
    provided_keys: set[str],
    cwd: str | os.PathLike[str],
) -> list[dict[str, str]]:
    """File-like bind mounts missing content and host file."""
    missing: list[dict[str, str]] = []
    for service, spec in services.items():
        for vol in spec.volumes or []:
            bind = parse_bind_mount(vol)
            if bind is None or not is_file_like_bind(bind.source):
                continue
            if normalize_rel(bind.source) in provided_keys:
                continue
            safe = resolve_safe(cwd, bind.source)
            if not safe.get("ok"):
                continue
            abs_path = str(safe["abs"])
            if not Path(abs_path).exists():
                missing.append({"service": service, "path": bind.source})
    return missing


def find_invalid_file_binds(
    services: dict[str, ServiceSpec],
    cwd: str | os.PathLike[str],
) -> list[dict[str, str]]:
    """Detect missing or directory-squatting file bind sources."""
    bad: list[dict[str, str]] = []
    for service, spec in services.items():
        for vol in spec.volumes or []:
            bind = parse_bind_mount(vol)
            if bind is None or not is_file_like_bind(bind.source):
                continue
            safe = resolve_safe(cwd, bind.source)
            if not safe.get("ok"):
                continue
            abs_path = Path(str(safe["abs"]))
            if not abs_path.exists():
                bad.append(
                    {"service": service, "path": bind.source, "reason": "missing"}
                )
            elif abs_path.is_dir():
                bad.append(
                    {"service": service, "path": bind.source, "reason": "directory"}
                )
    return bad


def stage_config_files(
    cwd: str | os.PathLike[str],
    services: dict[str, ServiceSpec],
    config_files: dict[str, str] | None,
) -> dict[str, object]:
    """Validate provided configFiles against file binds and size caps."""
    provided = config_files or {}
    file_binds: set[str] = set()
    for spec in services.values():
        for vol in spec.volumes or []:
            bind = parse_bind_mount(vol)
            if bind is not None and is_file_like_bind(bind.source):
                file_binds.add(normalize_rel(bind.source))

    total = 0
    staged: list[StagedConfigFile] = []
    for key, content in provided.items():
        safe = resolve_safe(cwd, key)
        if not safe.get("ok"):
            return {
                "ok": False,
                "error": f'unsafe config file path "{key}": {safe["error"]}',
            }
        norm = normalize_rel(key)
        if norm not in file_binds:
            return {
                "ok": False,
                "error": f'configFiles entry "{key}" matches no file bind mount',
            }
        nbytes = len(content.encode("utf-8"))
        if nbytes > MAX_FILE_BYTES:
            return {
                "ok": False,
                "error": f'config file "{key}" exceeds 64 KiB',
            }
        total += nbytes
        staged.append(StagedConfigFile(path=norm, content=content, bytes=nbytes))

    if total > MAX_TOTAL_BYTES:
        return {"ok": False, "error": "config files total exceeds 256 KiB"}
    return {"ok": True, "staged": staged}


def snapshot_config_files(
    cwd: str | os.PathLike[str], files: list[StagedConfigFile]
) -> list[ConfigFileSnapshot]:
    """Snapshot existing file content before staging writes."""
    snapshots: list[ConfigFileSnapshot] = []
    for file in files:
        abs_path = Path(cwd).resolve() / file.path
        existed = abs_path.is_file()
        previous = abs_path.read_text(encoding="utf-8") if existed else None
        snapshots.append(
            ConfigFileSnapshot(
                abs=str(abs_path),
                existed=existed,
                previous_content=previous,
            )
        )
    return snapshots


def write_config_files(
    cwd: str | os.PathLike[str], files: list[StagedConfigFile]
) -> None:
    """Write staged config files, recovering from directory squatters."""
    for file in files:
        safe = resolve_safe(cwd, file.path)
        if not safe.get("ok"):
            raise ValueError(f'refusing to write "{file.path}": {safe["error"]}')
        abs_path = Path(str(safe["abs"]))
        if abs_path.exists() and abs_path.is_dir():
            shutil.rmtree(abs_path)
        abs_path.parent.mkdir(parents=True, exist_ok=True)
        abs_path.write_text(file.content, encoding="utf-8")


def restore_config_files(snapshots: list[ConfigFileSnapshot]) -> None:
    """Restore config files from snapshots after a failed apply."""
    for snapshot in snapshots:
        path = Path(snapshot.abs)
        if snapshot.existed:
            path.write_text(snapshot.previous_content or "", encoding="utf-8")
        elif path.exists():
            if path.is_dir():
                shutil.rmtree(path)
            else:
                path.unlink()


__all__ = [
    "BindMount",
    "ConfigFileSnapshot",
    "StagedConfigFile",
    "detect_missing_config_files",
    "find_invalid_file_binds",
    "is_file_like_bind",
    "normalize_rel",
    "parse_bind_mount",
    "resolve_safe",
    "restore_config_files",
    "snapshot_config_files",
    "stage_config_files",
    "write_config_files",
]
